---
title: "MLML2R package User's Guide"
shorttitle: "MLML2R"
package: MLML2R
abstract: >
  We present a guide to the \R{} package \CRANpkg{MLML2R}. The package provides computational efficient maximum likelihood estimates of DNA methylation and hydroxymethylation proportions when data from the DNA processing methods bisulfite conversion (BS), oxidative bisulfite conversion (ox-BS), and Tet-assisted bisulfite conversion (TAB) are available. Estimates can be obtained when data from all the three methods are available or when any combination of only two of them are available. The package does not depend on other \R{} packages, allowing the user to read and preprocess the data with any given software, to import the results into \R{} in matrix format, to obtain the maximum likelihood 5-hmC and 5-mC estimates and use them as input for other packages traditionally used in genomic data analysis, such as \Biocpkg{minfi}, \Biocpkg{sva} and \Biocpkg{limma}.
bibliography: refs.bib
author: "Samara F. Kiihl and Maria Tellez-Plaza"
date: "`r Sys.Date()`"
output: 
  BiocStyle::pdf_document:
   fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Using MLML2R}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE,eval=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```


# Introduction


In a given CpG site from a single cell we will either have a C or a T after DNA processing conversion methods, with a different interpretation for each of the available methods. This is a binary outcome and we assume a Binomial model and use the maximum likelihood estimation method to obtain the estimates for hydroxymethylation and methylation proportions.

T reads are referred to as converted cytosine and C reads are referred to as unconverted cytosine. Conventionally, T counts are also referred to as unmethylated counts, and C counts as methylated counts. In case of Infinium Methylation arrays, we have intensities representing the methylated (M) and unmethylated (U) channels that are proportional to the number of unconverted and converted cytosines (C and T, respectively). The most used summary from these experiments is  the proportion $\beta=\frac{M}{M+U}$, commonly referred to as \textit{beta-value}, which reflects the methylation level at a CpG site. Naively using the difference between betas from BS and oxBS as an estimate of 5-hmC (hydroxymethylated cytosine), and the difference between betas from BS and TAB as an estimate of 5-mC (methylated cytosine) can many times provide negative proportions and instances where the sum of uC (unmodified cytosine), 5-mC and 5-hmC proportions is greater than one due.

\CRANpkg{MLML2R} package allows the user to jointly estimate hydroxymethylation and methylation consistently and efficiently.

The function \Rfunction{MLML} takes as input the data from the different methods and returns the estimated proportion of methylation, hydroxymethylation and unmethylation for a given CpG site. Table 1 presents the arguments of the \Rfunction{MLML} and Table 2 lists the results returned by the function.

The function assumes that the order of the rows and columns in the input matrices are consistent. In addition, all the input matrices must have the same dimension. Usually, rows represent CpG loci and columns are the samples.

Arguments           |       Description      
--------------------|----------------------------------------------------
\Robject{G.matrix}  | Unmethylated channel (Converted cytosines/ T counts) from  TAB-conversion (reflecting 5-C + 5-mC). 
\Robject{H.matrix}  | Methylated channel (Unconverted cytosines/ C counts) from TAB-conversion (reflecting True 5-hmC). 
\Robject{L.matrix}	| Unmethylated channel (Converted cytosines/ T counts) from oxBS-conversion (reflecting 5-C + 5-hmC). 
\Robject{M.matrix}	| Methylated channel (Unconverted cytosines/ C counts) from  oxBS-conversion (reflecting True 5-mC). 
\Robject{T.matrix}	| Methylated channel (Unconverted cytosines/ C counts) from  standard BS-conversion (reflecting 5-mC+5-hmC). 
\Robject{U.matrix}	| Unmethylated channel (Converted cytosines/ T counts) from standard BS-conversion (reflecting True 5-C). 
: `MLML` function and random variable notation.

Value | Description
------|------------------
\Robject{mC}	| maximum likelihood estimate for the proportion of methylation
\Robject{hmC}	| maximum likelihood estimate for the proportion of hydroxymethylation
\Robject{C}	| maximum likelihood estimate for the proportion of unmethylation
\Robject{methods}	| the conversion methods used to produce the MLE
: Results returned from the \Rfunction{MLML} function


# Worked examples

## Publicly available array data: oxBS and BS methods

We will use the dataset from @10.1371/journal.pone.0118202, which consists of eight DNA samples from the same DNA source treated with oxBS-BS and hybridized to the Infinium 450K array.

When data is obtained through Infinium Methylation arrays, we recommend the use of the \Biocpkg{minfi} package [@minfi], a well-established tool for reading, preprocessing and analysing DNA methylation data from these platforms. Although our example relies on \Biocpkg{minfi} and other \Bioconductor{} tools, \CRANpkg{MLML2R} does not depend on any packages. Thus, the user is free to read and preprocess the data using any software of preference and then import the intensities (or T and C counts) for the methylated and unmethylated channel (or converted and uncoverted cytosines) into \R{} in matrix format.

To start this example we will need the following packages:
```{r packages1,eval=TRUE,message=FALSE,warning=FALSE}
library(MLML2R)
library(minfi)
library(GEOquery)
library(IlluminaHumanMethylation450kmanifest)
```


It is usually best practice to start the analysis from the raw data, which in the case of the 450K array is a \verb|.IDAT| file.

The raw files are deposited in GEO and can be downloaded by using the \Rfunction{getGEOSuppFiles}. There are two files for each replicate, since the 450k array is a two-color array. The \verb|.IDAT| files are downloaded in compressed format and need to be uncompressed before they are read by the \Rfunction{read.metharray.exp} function.

```{r getData1,eval=FALSE}
getGEOSuppFiles("GSE63179")
untar("GSE63179/GSE63179_RAW.tar", exdir = "GSE63179/idat")

list.files("GSE63179/idat", pattern = "idat")
files <- list.files("GSE63179/idat", pattern = "idat.gz$", full = TRUE)
sapply(files, gunzip, overwrite = TRUE)
```

The \verb|.IDAT| files can now be read:
```{r readData1,eval=FALSE}
rgSet <- read.metharray.exp("GSE63179/idat")
```

To access phenotype data we use the \Rfunction{pData} function. The phenotype data is not yet available from the \Robject{rgSet}.
```{r,eval=FALSE}
pData(rgSet)
```
In this example the phenotype is not really relevant, since we have only one sample: male, 25 years old. What we do need is the information about the conversion method used in each replicate: BS or oxBS. We will access this information automatically from GEO:
```{r getPheno1,eval=FALSE}
if (!file.exists("GSE63179/GSE63179_series_matrix.txt.gz"))
download.file(
"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE63nnn/GSE63179/matrix/GSE63179_series_matrix.txt.gz",
"GSE63179/GSE63179_series_matrix.txt.gz")

geoMat <- getGEO(filename="GSE63179/GSE63179_series_matrix.txt.gz",getGPL=FALSE)
pD.all <- pData(geoMat)

#Another option
#geoMat <- getGEO("GSE63179")
#pD.all <- pData(geoMat[[1]])

pD <- pD.all[, c("title", "geo_accession", "characteristics_ch1.1",
                 "characteristics_ch1.2","characteristics_ch1.3")]
pD
```


This phenotype data needs to be merged into the methylation data. The following commands guarantee we have the same replicate identifier in both datasets before merging.

```{r,eval=FALSE}
sampleNames(rgSet) <- sapply(sampleNames(rgSet),function(x)
  strsplit(x,"_")[[1]][1])
rownames(pD) <- pD$geo_accession
pD <- pD[sampleNames(rgSet),]
pData(rgSet) <- as(pD,"DataFrame")
rgSet
```

The \Robject{rgSet} is an object from \Rclass{RGChannelSet} class used for two color data (green and red channels). The input in the \Rfunction{MLML} function are matrices with methylated and unmethylated information from each conversion method. We can use the \Rclass{MethylSet} class, which contains the methylated and unmethylated signals. The most basic way to construct a \Rclass{MethylSet} is  using the function \Rfunction{preprocessRaw}. Here we chose the function \Rfunction{preprocessNoob} [@noob] for background correction, dye bias normalization and construction of the \Rclass{MethylSet}.


```{r Preprocess1,eval=FALSE}
MSet.noob<- preprocessNoob(rgSet)
```


After the preprocessed steps we can use \Rfunction{MLML} from the \CRANpkg{MLML2R} package.


The BS replicates are in columns 1, 3, 5, and 6 (information from `pD$title`). The remaining columns are from the oxBS treated replicates.

```{r,eval=FALSE}
MethylatedBS <- getMeth(MSet.noob)[,c(1,3,5,6)]
UnMethylatedBS <- getUnmeth(MSet.noob)[,c(1,3,5,6)]
MethylatedOxBS <- getMeth(MSet.noob)[,c(7,8,2,4)]
UnMethylatedOxBS <- getUnmeth(MSet.noob)[,c(7,8,2,4)]
```


 When only two methods are available, the default option of \Rfunction{MLML} function returns the exact constrained maximum likelihood estimates  using the the pool-adjacent-violators algorithm (PAVA) [@ayer1955].
 

```{r MLML2Rexact1,eval=FALSE}
results_exact <- MLML(T.matrix = MethylatedBS , U.matrix = UnMethylatedBS,
                      L.matrix = UnMethylatedOxBS, M.matrix = MethylatedOxBS)
```



Maximum likelihood estimate via EM-algorithm approach [@Qu:MLML] is obtained with the option \verb|iterative=TRUE|. In this case, the default (or user specified) \verb|tol| is considered in the iterative method.


```{r MLML2REM1,eval=FALSE}
results_em <- MLML(T.matrix = MethylatedBS , U.matrix = UnMethylatedBS,
                   L.matrix = UnMethylatedOxBS, M.matrix = MethylatedOxBS,
                   iterative = TRUE)
```

The estimates are very similar for both methods:
```{r,echo=TRUE,eval=FALSE}
all.equal(results_exact$hmC,results_em$hmC,scale=1)
```



```{r plot,echo=FALSE,eval=FALSE}
pdf(file="Real1_estimates.pdf",width=15,height=5)
par(mfrow =c(1,3))
densityPlot(results_exact$hmC,main= "5-hmC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion")
densityPlot(results_exact$mC,main= "5-mC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion")
densityPlot(results_exact$C,main= "uC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion")
dev.off()
```

```{r,echo=FALSE,fig.width=15,fig.height=5,fig.cap="Estimated proportions of hydroxymethylation, methylation and unmethylation for the CpGs in the dataset using the MLML function with default options."}
knitr::include_graphics("Real1_estimates.pdf") 
```

## Publicly available array data: TAB and BS methods

We will use the dataset from @Thienpont2016, which consists of 24 DNA samples treated with TAB-BS and hybridized to the Infinium 450K array from newly diagnosed and untreated non-small-cell lung cancer patients (12 normoxic and 12 hypoxic tumours). The dataset is deposited under GEO accession number [GSE71398](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71398).

We will need the following packages:
```{r packages2,eval=TRUE,message=FALSE,warning=FALSE}
library(MLML2R)
library(minfi)
library(GEOquery)
library(IlluminaHumanMethylation450kmanifest)
library(wateRmelon)
```

Obtaining the data:
```{r getData2,eval=FALSE}
getGEOSuppFiles("GSE71398")
untar("GSE71398/GSE71398_RAW.tar", exdir = "GSE71398/idat")

list.files("GSE71398/idat", pattern = "idat")
files <- list.files("GSE71398/idat", pattern = "idat.gz$", full = TRUE)
sapply(files, gunzip, overwrite = TRUE)
```

Reading the \verb|.IDAT| files:
```{r readData2,eval=FALSE}
rgSet <- read.metharray.exp("GSE71398/idat")
```



The phenotype data is not yet available from the \Robject{rgSet}.
```{r,eval=FALSE}
pData(rgSet)
```

We need to correctly identify the 24 DNA samples: 12 normoxic and 12 hypoxic non-small-cell lung cancer. We also need the information about the conversion method used in each replicate: BS or TAB. We will access this information automatically from GEO:

```{r getPheno2,eval=FALSE}
if (!file.exists("GSE71398/GSE71398_series_matrix.txt.gz"))
download.file(
"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE71nnn/GSE71398/matrix/GSE71398_series_matrix.txt.gz",
"GSE71398/GSE71398_series_matrix.txt.gz")

geoMat <- getGEO(filename="GSE71398/GSE71398_series_matrix.txt.gz",getGPL=FALSE)
pD.all <- pData(geoMat)

#Another option
#geoMat <- getGEO("GSE71398")
#pD.all <- pData(geoMat[[1]])

pD <- pD.all[, c("title", "geo_accession", "source_name_ch1",
                 "tabchip or bschip:ch1","hypoxia status:ch1",
                 "tumor name:ch1","batch:ch1","platform_id")]
pD$method <- pD$`tabchip or bschip:ch1`
pD$group <- pD$`hypoxia status:ch1`
pD$sample <- pD$`tumor name:ch1`
pD$batch <- pD$`batch:ch1`
```


This phenotype data needs to be merged into the methylation data. The following commands guarantee we have the same replicate identifier in both datasets before merging.

```{r,eval=FALSE}
sampleNames(rgSet) <- sapply(sampleNames(rgSet),function(x)   
  strsplit(x,"_")[[1]][1])
rownames(pD) <- as.character(pD$geo_accession)
pD <- pD[sampleNames(rgSet),]
pData(rgSet) <- as(pD,"DataFrame")
rgSet
```


The input in the \Rfunction{MLML} function accepts as input a \Rclass{MethylSet}, which contains the methylated and unmethylated signals. We chose the function \Rfunction{preprocessNoob} [@noob] for background correction, dye-bias normalization and construction of the \Rclass{MethylSet}. In addition, the function \Rfunction{BMIQ} [@Teschendorff2012] from the package \Biocpkg{wateRmelon} [@Pidsley2013] was used for probe-type bias correction. A discussion of this preprocessing procedure is presented by @Liu2016.


```{r preprocess2,eval=FALSE}
## Noob
MSet.noob<- preprocessNoob(rgSet)

BSindex <- which(pD$method=="BSchip")
TABindex <- which(pD$method=="TABchip")  

## BMIQ
anno <- getAnnotation(MSet.noob)
beta.b <- getBeta(MSet.noob, type = "Illumina")
design.v <- as.vector(anno$Type)
design.v[design.v == "I"] = 1
design.v[design.v == "II"] = 2
design.v <- as.numeric(design.v)
coln = colnames(beta.b)

beta.noob.bmiq <- BMIQ(beta.b, design.v = design.v,sampleID = 1:48)

beta_BS <- beta.noob.bmiq$nbeta[,BSindex]
beta_TAB <- beta.noob.bmiq$nbeta[,TABindex]

# Total Signal = methylated + unmethylated
TotalBS <- getMeth(MSet.noob[,BSindex]) + getUnmeth(MSet.noob[,BSindex])
TotalTAB <- getMeth(MSet.noob[,TABindex]) + getUnmeth(MSet.noob[,TABindex])

MethylatedBS <- beta_BS*TotalBS
UnMethylatedBS <- (1-beta_BS)*TotalBS
MethylatedTAB <- beta_TAB*TotalTAB
UnMethylatedTAB <- (1-beta_TAB)*TotalTAB
```


We can now use \Rfunction{MLML} from the \CRANpkg{MLML2R} package.

One needs to carefully check if the columns across the different input matrices represent the same sample. In this example, all matrices have the samples consistently represented in the columns: sample 1 in the first column, sample 2 in the second, and so forth.


 When only two methods are available, the default option of \Rfunction{MLML} function returns the exact constrained maximum likelihood estimates  using the the pool-adjacent-violators algorithm (PAVA) [@ayer1955].

```{r,eval=FALSE}
results_exact <- MLML(T.matrix = MethylatedBS , U.matrix = UnMethylatedBS,
                      G.matrix = UnMethylatedTAB, H.matrix = MethylatedTAB)
```


Maximum likelihood estimate via EM-algorithm approach [@Qu:MLML] is obtained with the option \verb|iterative=TRUE|. In this case, the default (or user specified) \verb|tol| is considered in the iterative method.


```{r MLML2REM2,eval=FALSE}
results_em <- MLML(T.matrix = MethylatedBS , U.matrix = UnMethylatedBS,
                   G.matrix = UnMethylatedTAB, H.matrix = MethylatedTAB,
                   iterative = TRUE)
```

The estimates for 5-hmC proportions are very similar for both methods:
```{r,echo=TRUE,eval=FALSE}
all.equal(results_exact$hmC,results_em$hmC,scale=1)
```

The estimates for 5-mC proportions are very similar for both methods:
```{r,echo=TRUE,eval=FALSE}
all.equal(results_exact$mC,results_em$mC,scale=1)
```



```{r plot3,echo=FALSE,eval=FALSE}
pdf(file="Real2a_estimates.pdf",width=15,height=5)
par(mfrow =c(1,3))
densityPlot(results_exact$hmC,main= "5-hmC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion",sampGroups=pD$group[BSindex])
densityPlot(results_exact$mC,main= "5-mC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion",sampGroups=pD$group[BSindex])
densityPlot(results_exact$C,main= "uC estimates - PAVA",cex.axis=1.4,cex.main=1.5,cex.lab=1.4,xlab="Proportion",sampGroups=pD$group[BSindex])
dev.off()
```

```{r,echo=FALSE,fig.width=15,fig.height=5,fig.cap="Estimated proportions of hydroxymethylation, methylation and unmethylation for the CpGs in the dataset using the MLML function with default options."}
knitr::include_graphics("Real2a_estimates.pdf") 
```


## Simulated data


To illustrate the package when all the three methods are available or when any combination of only two of them are available, we will simulate a dataset. 

We will use a sample of the estimates of 5-mC, 5-hmC and uC of the previous oxBS+BS example as the true proportions, as shown in Figure 3.



Two replicate samples with 1000 CpGs will be simulated. For CpG $i$ in sample $j$:

$$T_{i,j} \sim Binomial(n=c_{i,j},p=p_m+p_h)$$
$$M_{i,j} \sim Binomial(n=c_{i,j}, p=p_m)$$
$$H_{i,j} \sim Binomial(n=c_{i,j},p=p_h)$$
$$U_{i,j}=c_{i,j}-T_{i,j}$$
$$L_{i,j}=c_{i,j}-M_{i,j}$$
$$G_{i,j}=c_{i,j}-H_{i,j}$$
where the random variables are defined in Table 1, and $c_{i,j}$ represents the coverage for CpG $i$ in sample $j$.



The following code produce the simulated data:

```{r simulation,echo=TRUE,eval=FALSE}
set.seed(112017)

index <- sample(1:dim(results_exact$mC)[1],1000,replace=FALSE) # 1000 CpGs

Coverage <- round(MethylatedBS+UnMethylatedBS)[index,1:2] # considering 2 samples

temp1 <- data.frame(n=as.vector(Coverage),
                    p_m=c(results_exact$mC[index,1],
                          results_exact$mC[index,1]),
                    p_h=c(results_exact$hmC[index,1],
                          results_exact$hmC[index,1]))

MethylatedBS_temp <- c()
for (i in 1:dim(temp1)[1])
{
  MethylatedBS_temp[i] <- rbinom(n=1, size=temp1$n[i], 
                                 prob=(temp1$p_m[i]+temp1$p_h[i]))
}


UnMethylatedBS_sim2 <- matrix(Coverage - MethylatedBS_temp,ncol=2)
MethylatedBS_sim2 <- matrix(MethylatedBS_temp,ncol=2)


MethylatedOxBS_temp <- c()
for (i in 1:dim(temp1)[1])
{
  MethylatedOxBS_temp[i] <- rbinom(n=1, size=temp1$n[i], prob=temp1$p_m[i])
}

UnMethylatedOxBS_sim2 <- matrix(Coverage - MethylatedOxBS_temp,ncol=2)
MethylatedOxBS_sim2 <- matrix(MethylatedOxBS_temp,ncol=2)


MethylatedTAB_temp <- c()
for (i in 1:dim(temp1)[1])
{
  MethylatedTAB_temp[i] <- rbinom(n=1, size=temp1$n[i], prob=temp1$p_h[i])
}


UnMethylatedTAB_sim2 <- matrix(Coverage - MethylatedTAB_temp,ncol=2)
MethylatedTAB_sim2 <- matrix(MethylatedTAB_temp,ncol=2)

true_parameters_sim2 <- data.frame(p_m=results_exact$mC[index,1],
                                   p_h=results_exact$hmC[index,1])
true_parameters_sim2$p_u <- 1-true_parameters_sim2$p_m-true_parameters_sim2$p_h
```


```{r eval=FALSE,echo=FALSE}
save(true_parameters_sim2,MethylatedBS_sim2,UnMethylatedBS_sim2,MethylatedOxBS_sim2,UnMethylatedOxBS_sim2,MethylatedTAB_sim2,UnMethylatedTAB_sim2,file="Data_sim2.rds")
```


```{r plot2,echo=FALSE,eval=FALSE}
pdf(file="True_parameters.pdf",width=15,height=5)
par(mfrow =c(1,3))
plot(density(results_exact$hmC[index,1]),main= "True 5-hmC",xlab="Proportions",xlim=c(0,1),ylim=c(0,10),cex.axis=1.5,cex.main=1.5,cex.lab=1.5)
plot(density(results_exact$mC[index,1]),main= "True 5-mC",xlab="Proportions",xlim=c(0,1),ylim=c(0,10),cex.axis=1.5,cex.main=1.5,cex.lab=1.5)
plot(density(results_exact$C[index,1]),main= "True uC",ylim=c(0,10),xlab="Proportions",xlim=c(0,1),cex.axis=1.5,cex.main=1.5,cex.lab=1.5)
dev.off()
```

```{r,echo=FALSE,fig.width=15,fig.height=5,fig.cap="True proportions of hydroxymethylation, methylation and unmethylation for the CpGs used to generate the datasets."}
knitr::include_graphics("True_parameters.pdf") 
```



### BS and oxBS methods


```{r,echo=FALSE,eval=TRUE}
load("Data_sim2.rds")
```

When only two methods are available, the default option returns the exact constrained maximum likelihood estimates  using the the pool-adjacent-violators algorithm (PAVA) [@ayer1955].


```{r}
library(MLML2R)
 results_exactBO1 <- MLML(T.matrix = MethylatedBS_sim2 , 
                          U.matrix = UnMethylatedBS_sim2,
                          L.matrix = UnMethylatedOxBS_sim2, 
                          M.matrix = MethylatedOxBS_sim2)
```

Maximum likelihood estimate via EM-algorithm approach [@Qu:MLML] is obtained with the option \verb|iterative=TRUE|. In this case, the default (or user specified) \verb|tol| is considered in the iterative method.

```{r}
 results_emBO1 <- MLML(T.matrix = MethylatedBS_sim2, 
                       U.matrix = UnMethylatedBS_sim2,
                       L.matrix = UnMethylatedOxBS_sim2, 
                       M.matrix = MethylatedOxBS_sim2,
                       iterative=TRUE)
```

 When only two methods are available, we highly recommend the default option \Rcode{iterative=FALSE} since the difference in the estimates obtained via EM and exact constrained is very small, but the former requires more computational effort:

```{r}
 all.equal(results_emBO1$hmC,results_exactBO1$hmC,scale=1)
```

```{r}
 library(microbenchmark)
 mbmBO1 = microbenchmark(
    EXACT = MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2),
    EM =    MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2,
                 iterative=TRUE),
    times=10)
 mbmBO1
```


Comparison between approximate exact constrained and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_exactBO1$hmC[,1],scale=1)
```

Comparison between EM-algorithm and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_emBO1$hmC[,1],scale=1)
```

### BS and TAB methods

Using PAVA:
```{r}
results_exactBT1 <- MLML(T.matrix = MethylatedBS_sim2, 
                         U.matrix = UnMethylatedBS_sim2,
                         G.matrix = UnMethylatedTAB_sim2, 
                         H.matrix = MethylatedTAB_sim2)
```

Using EM-algorithm:
```{r}
 results_emBT1 <- MLML(T.matrix = MethylatedBS_sim2, 
                       U.matrix = UnMethylatedBS_sim2,
                       G.matrix = UnMethylatedTAB_sim2, 
                       H.matrix = MethylatedTAB_sim2,
                       iterative=TRUE)
```

Comparison between PAVA and EM:

```{r}
 all.equal(results_emBT1$hmC,results_exactBT1$hmC,scale=1)
```

```{r}
 mbmBT1 = microbenchmark(
    EXACT = MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2),
    EM =    MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2,
                 iterative=TRUE),
    times=10)
 mbmBT1
```


Comparison between approximate exact constrained and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_exactBT1$hmC[,1],scale=1)
```

Comparison between EM-algorithm and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_emBT1$hmC[,1],scale=1)
```

### oxBS and TAB methods

 Using PAVA:
```{r}
 results_exactOT1 <- MLML(L.matrix = UnMethylatedOxBS_sim2, 
                          M.matrix = MethylatedOxBS_sim2,
                          G.matrix = UnMethylatedTAB_sim2, 
                          H.matrix = MethylatedTAB_sim2)
```

 Using EM-algorithm:
```{r}
 results_emOT1 <- MLML(L.matrix = UnMethylatedOxBS_sim2, 
                       M.matrix = MethylatedOxBS_sim2,
                       G.matrix = UnMethylatedTAB_sim2, 
                       H.matrix = MethylatedTAB_sim2,
                       iterative=TRUE)
```

Comparison between PAVA and EM:

```{r}
 all.equal(results_emOT1$hmC,results_exactOT1$hmC,scale=1)
```

```{r}
 mbmOT1 = microbenchmark(
    EXACT = MLML(L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2),
    EM =    MLML(L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2,
                 iterative=TRUE),
    times=10)
 mbmOT1
```

Comparison between approximate exact constrained and true 5-hmC proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_exactOT1$hmC[,1],scale=1)
```

Comparison between EM-algorithm and true 5-hmC proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_emOT1$hmC[,1],scale=1)
```


### BS, oxBS and TAB methods

When data from the three methods are available, the default otion in the \Rfunction{MLML} function returns the constrained maximum likelihood estimates using an approximated solution for Lagrange multipliers method.


```{r}

results_exactBOT1 <- MLML(T.matrix = MethylatedBS_sim2, 
                          U.matrix = UnMethylatedBS_sim2,
                          L.matrix = UnMethylatedOxBS_sim2, 
                          M.matrix = MethylatedOxBS_sim2,
                          G.matrix = UnMethylatedTAB_sim2, 
                          H.matrix = MethylatedTAB_sim2)
```

Maximum likelihood estimate via EM-algorithm approach [@Qu:MLML] is obtained with the option \verb|iterative=TRUE|. In this case, the default (or user specified) \verb|tol| is considered in the iterative method.

```{r}
 results_emBOT1 <- MLML(T.matrix = MethylatedBS_sim2, 
                        U.matrix = UnMethylatedBS_sim2,
                        L.matrix = UnMethylatedOxBS_sim2, 
                        M.matrix = MethylatedOxBS_sim2,
                        G.matrix = UnMethylatedTAB_sim2, 
                        H.matrix = MethylatedTAB_sim2,iterative=TRUE)
```


We recommend the default option \Rcode{iterative=FALSE} since the difference in the estimates obtained via EM and the approximate exact constrained is very small, but the former requires more computational effort:

```{r}
 all.equal(results_emBOT1$hmC,results_exactBOT1$hmC,scale=1)
```


```{r computationCost}
 mbmBOT1 = microbenchmark(
    EXACT = MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2),
    EM =    MLML(T.matrix = MethylatedBS_sim2, 
                 U.matrix = UnMethylatedBS_sim2,
                 L.matrix = UnMethylatedOxBS_sim2, 
                 M.matrix = MethylatedOxBS_sim2,
                 G.matrix = UnMethylatedTAB_sim2, 
                 H.matrix = MethylatedTAB_sim2,
                 iterative=TRUE),
    times=10)
 mbmBOT1
```


Comparison between approximate exact constrained and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_exactBOT1$hmC[,1],scale=1)
```

Comparison between EM-algorithm and true hydroxymethylation proportion used in simulation:
```{r}
all.equal(true_parameters_sim2$p_h,results_emBOT1$hmC[,1],scale=1)
```


# References
